### ISSUE TITLE
RangeError: Maximum call stack size in getTypeAtLocation in files importing in series with template literals

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [microsoft/TypeScript#59255]
### 🔎 Search Terms

getTypeAtLocation getTypeOfNode checkNonNullExpression

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about type checking and APIs

### ⏯ Playground Link

https://github.com/JoshuaKGoldberg/repros/blob/typescript-eslint-rangeerror-170

### 💻 Code

Files are set up in a singly linked list of `export`s and `import`s.
Within each case:

- `src/index.ts` imports a `pipe0` function from `src/pipe0.ts`
- `src/pipe0.ts` imports a `pipe1` function from `src/pipe1.ts`
- ...
- The last pipe file `N` exports a `pipe${N}` function

[typescript-eslint's typed linting](https://typescript-eslint.io/getting-started/typed-linting) is then run on the `src/` directory.

### 🙂 Expected behavior

Types should be retrieved successfully at scales of 100+ files importing in a row _and_ using template literal expressions.

### 🙁 Actual behavior

```plaintext
Oops! Something went wrong! :(

ESLint: 9.6.0

RangeError: Maximum call stack size exceeded
Occurred while linting /repros/cases/files-170-types-project/src/pipe0.ts:5
Rule: "@typescript-eslint/no-unsafe-return"
    at checkTemplateExpression (/repros/node_modules/typescript/lib/typescript.js:82916:11)
    at checkExpressionWorker (/repros/node_modules/typescript/lib/typescript.js:83375:16)
    at checkExpression (/repros/node_modules/typescript/lib/typescript.js:83303:32)
    at checkNonNullExpression (/repros/node_modules/typescript/lib/typescript.js:77538:29)
    at checkPropertyAccessExpression (/repros/node_modules/typescript/lib/typescript.js:77617:162)
    at checkExpressionWorker (/repros/node_modules/typescript/lib/typescript.js:83383:16)
    at checkExpression (/repros/node_modules/typescript/lib/typescript.js:83303:32)
    at resolveCallExpression (/repros/node_modules/typescript/lib/typescript.js:79548:20)
    at resolveSignature (/repros/node_modules/typescript/lib/typescript.js:79997:16)
    at getResolvedSignature (/repros/node_modules/typescript/lib/typescript.js:80023:18)
    ...
```

### Additional information about the issue

Running `tsc` on the CLI works just fine at this scale. It's the type checker API that's crashing.

https://github.com/JoshuaKGoldberg/repros/blob/typescript-eslint-rangeerror-170/README.md#full-error-stack contains the full error stack. Looking at the last lines in it (the first functions called):

* The first TypeScript API called is `getTypeAtLocation`, and it only shows up once
* The next few APIs called are also only called once: `getTypeOfNode`, `getRegularTypeOfExpression`, `checkNonNullExpression`
* `checkExpression` is called around 875+ times

Also copying from https://github.com/JoshuaKGoldberg/repros/blob/typescript-eslint-rangeerror-170/README.md#testing-notes:
* Tested on TypeScript 4.7.3, 5.1.3, and 5.3.3: all reproduce the issue.
* Avoiding template literal strings fixes the errors.

**Comment by weswigham** [comment]
I thought it was maybe something yarn related when I couldn't repro this, so redid the testing with yarn, but no dice. Am I missing something? I've been running the repro code, and I can't get it to repro a crash under any circumstance? With or without `yarn`, on node `22.3.0`. It's a bit slow, but I get a big
![Screenshot 2024-08-06 134426](https://github.com/user-attachments/assets/a543d1e6-8ef1-47e8-8398-8d72052823c3)
when yarn via `yarn` and `yarn comparison`. I even modified the versions in the `index.ts` script and `package.json` to explicitly use TS 5.5.3, and then 5.3.3 - neither of those got me an error, either. I also wrote a script to literally call `getTypeAtLocation` on the node at the supposedly crashing location in the README, and that, too, does not crash. Ran it inside a codespace on a linux host on node `22.6.0` - also no crash.

**Comment by JoshuaKGoldberg** [comment]
Hmm, I still reproduce this locally. Maybe the 170 is just coincidentally what I get on my machine's memory allotment? If you try switching to `values: [300]`, or even `values: [500]`, does that cause a crash for you?

I pushed https://github.com/JoshuaKGoldberg/repros/commit/4d0c3a125be97e32b7ad3550e7908823bec9465f that adjusted values to go from `100` to `300`.

**Comment by weswigham** [comment]
Alright, yeah, there's one at a higher number of files in series
![image](https://github.com/user-attachments/assets/92caaeab-a22d-4331-aecf-4b214954465b)
but the project and services results look the same. And, perhaps moreover, when I set the `NON_TEMPLATE_STRINGS` environment variable and rerun, I _also_ get some failures at around the same point
![image](https://github.com/user-attachments/assets/63851649-d172-4ebd-9b6f-4dfc4cfbb703)

which looks like pretty similar behavior to me (the template literal case is going to use an extra two or so stack frames per file, so it'll just run out more reliably a bit sooner - the expression is just a bit more complex). As we're currently architected, we're *gonna* overflow at *some point* (though you can manually raise the max stack limit with a v8 flag), and I don't really see much difference between the two cases? Or is the point that we overflow at any point with either, even though a simple `tsc` on the commandline doesn't (which I'm guessing is just the result of a bottom-up check order instead of a top-down one)?

**Comment by JoshuaKGoldberg** [comment]
> the point that we overflow at any point with either

Exactly, yeah. I just included both APIs to show that both (either) will do it. The intent here is that regardless of the project API/service used, `getTypeAtLocation()` in this situation causes the error.

**Comment by weswigham** [comment]
Hm, well, I was thinking I could do something to make the template literal case shortcut a bit more, but it's not really the issue here - the issue is really just that the checker doesn't trampoline its' calls to schedule work and relies entirely on the callstack to track state. The commandline checks in a different order that happens to not start in the `index.ts` file, so gets to bottom out faster, cache the work done, then finish the rest. Technically, you can force the API into working the same way (likely more reliably) by organizing your `getTypeAtLocation` calls to operate on leaf node files first and then work your way up the topological graph - though that is still a workaround.

**Comment by RyanCavanaugh** [comment]
`getTypeAtLocation` is part of the pre-TypeScript-7 JavaScript Compiler API, which is no longer being developed. The reported stack overflow occurs when that API recursively checks a sufficiently long import chain; template literal expressions make the stack limit easier to reach but are not required for the failure.
